iceoryx_dust/test/moduletests/test_container_fixed_position_container.cpp

```diff
@@ -92,6 +94,326 @@ TEST_F(FixedPositionContainer_test, Capacity)
     EXPECT_THAT(sut.capacity(), Eq(CAPACITY));
 }
 
+// BEGIN test copy constructor
+
+TEST_F(FixedPositionContainer_test, UsingCopyCtorEmptyContainerResultsInEmptyContainer)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6c528ef3-9c2d-4eb2-93a9-2d998d0db380");
+
+    SutComplex copy_sut_complex{sut_complex};
+
+    EXPECT_THAT(copy_sut_complex.empty(), Eq(true));
+}
+
+TEST_F(FixedPositionContainer_test, UsingCopyCtorSingleElementContainerPreservesElement)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f3aaf452-77fa-4535-bf0b-37bedefc2bf6");
+
+    constexpr DataType EXPECTED_VALUE{42U};
+    constexpr SutComplex::IndexType EXPECTED_SIZE{1U};
+
+    sut_complex.emplace(EXPECTED_VALUE);
+    SutComplex copy_sut_complex{sut_complex};
+
+    EXPECT_THAT(copy_sut_complex.size(), Eq(EXPECTED_SIZE));
+
+    // actually call copyAssignment
+    EXPECT_THAT(copy_sut_complex.begin()->stats.copyAssignment, Eq(EXPECTED_SIZE));
+    EXPECT_THAT(copy_sut_complex.begin()->value, Eq(EXPECTED_VALUE));
+}
+
+TEST_F(FixedPositionContainer_test, UsingCopyCtorMultipleElementsContainerPreservesAllElements)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "6261f53e-8089-4b9b-9b2d-9da0016a2f1e");
+
+    constexpr SutComplex::IndexType EXPECTED_SIZE{4U};
+    std::vector<DataType> EXPECTED_VALUE = {0U, 1U, 2U, 3U};
+    for (SutComplex::IndexType i = 0; i < EXPECTED_SIZE; ++i)
+    {
+        sut_complex.emplace(EXPECTED_VALUE[i]);
+    }
+
+    SutComplex copy_sut_complex{sut_complex};
+
+    EXPECT_THAT(copy_sut_complex.size(), Eq(EXPECTED_SIZE));
+
+    // actually call copyAssignment
+    EXPECT_THAT(copy_sut_complex.begin()->stats.copyAssignment, Eq(EXPECTED_SIZE));
+    for (Sut::IndexType i = 0; i < EXPECTED_SIZE; ++i)
+    {
+        EXPECT_THAT(copy_sut_complex.iter_from_index(i)->value, Eq(EXPECTED_VALUE[i]));
+    }
+}
+
+TEST_F(FixedPositionContainer_test, UsingCopyCtorFullCapacityContainerPreservesAllElements)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "028704df-b2f3-4133-9c16-b9d2c6a79916");
+
+    fillSutComplex();
+    constexpr SutComplex::IndexType EXPECTED_SIZE{CAPACITY};
+
+    SutComplex copy_sut_complex{sut_complex};
+
+    EXPECT_THAT(copy_sut_complex.full(), Eq(true));
+
+    // actually call copyAssignment
+    EXPECT_THAT(copy_sut_complex.begin()->stats.copyAssignment, Eq(EXPECTED_SIZE));
+    for (Sut::IndexType i = 0; i < EXPECTED_SIZE; ++i)
+    {
+        EXPECT_THAT(copy_sut_complex.iter_from_index(i)->value, Eq(sut_complex.iter_from_index(i)->value));
+        EXPECT_THAT(copy_sut_complex.iter_from_index(i), Ne(sut_complex.iter_from_index(i)));
+    }
+}
+
+// END test copy constructor
```

Comment:
There are some missing tests:
- copy empty  container to non empty one
- copy non-empty container with `size` being less than, equal and greater than than the destination container
- copy non-empty container with `Index::FIRST` erased to an empty one
- copy non-empty container with `Index::FIRST` erased to non empty one
- copy non-empty container with `Index::FIRST` erased to non empty one with `Index::FIRST` erased
- copy non-empty container with `Index::FIRST` erased to non empty one with index `Index::FIRST` and `Index::FIRST + 1` erased
- copy non-empty container with `Index::FIRST`, `Index::LAST / 2` and `Index::LAST` erased to an empty one
- copy non-empty container with `Index::FIRST`, `Index::LAST / 2` and `Index::LAST` erased to a non empty one
- copy non-empty container with `Index::FIRST`, `Index::LAST / 2` and `Index::LAST` erased to a non empty one with `Index::FIRST` erased
- copy non-empty container with `Index::FIRST`, `Index::LAST / 2` and `Index::LAST` erased to a non empty one with `Index::FIRST + 1`,  `Index::LAST / 2 - 1` and `Index::LAST - 1` erased
- copy non-empty container with `Index::LAST` erased to full container but with `Index::FIRST` erased
- add a test to check that the source container did not change (just in case someone does a const_cast)

This also need to be added to the other move ctor and assignment operator tests.

You can have a look at https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md on how to come up with all these test cases :)

Reply:
I think I get it now :O

Reply:
Some of the tests mentioned above may be applicable only to assignment operations. For instance, I believe the "empty container to non-empty one" scenario can't be accomplished using a constructor; it can only be done via assignment.

Here's how the code comment might look, revised for clarity:

```c
SutComplex tmp;
tmp.emplace(rand_num);

// I believe this can only be accomplished through assignment
tmp = sut_complex;

EXPECT_THAT(tmp.empty(), Eq(true));
// ...
```

Please correct me if I'm wrong, otherwise I'm inclined to implement tests only where necessary :)

Reply:
Yes, that's correct. I think it would be even fine to have just one copy/move ctor test since they are implemented by delegating to the corresponding assignment operators anyway.

Reply:
I think I have on additional test. Could you please add a test where a partially filled container is moved or copied and then add new elements to the moved-to container till it is full. The moved-to container should also not be empty but partially filled and the slot and `Index::LAST` should also be occupied. This should ensure that the free list is not broken after the move/copy.

Reply:
You mean something like:
```c
// arrange: add some elements
constexpr DataType offset{32U};
sut_complex.emplace(offset);
sut_complex.emplace(offset+1);

// action:
Sutcomplex copy_or_move_sut_complex;
copy_or_move_sut_complex = sut_complex;

// assert: 
EXPECT_THAT(copy_or_move_sut_complex.full(), Eq(false));
EXPECT_THAT(copy_or_move_sut_complex.empty(), Eq(false));
EXPECT_THAT(copy_or_move_sut_complex.size(), Eq(2U));

// action: fill rest slots
for(IndexType i = 2U; i < CAPACITY; ++i)
{
    copy_or_move_sut_complex.emplace(offset+i);
}

// assert:
EXPECT_THAT(copy_or_move_sut_complex.full(), Eq(true));
EXPECT_THAT(copy_or_move_sut_complex.iter_from_index(SutComplex::Index::LAST)->value,
                        offset + CAPACITY - 1);

```
If something goes wrong, please tell me:)
If everything is fine, I'll append it to the todo list of test cases.

Reply:
More like
```cpp
SutComplex  sut2;
fillComplex(sut2); // aaargh, should have called this fill_complex()

sut2.erase(Sut::Index::FIRST);
sut2.erase(Sut::Index::LAST / 2U);

// do something similar with sut_complex; important is only that it is neither empty nor full

sut2 = sut_complex;

// now insert two elements and the third insertion should fail
// sut2 should also be full now and the size should equal the capacity
```
Something similar to above.

It is not necessary to do some asserts after the copy since that scenario is already tested. The new scenario is the filling of the destination container after the copy and only for this the checks need to be done.

Btw, there is the iceoryx developer meetup tomorrow around this time. Your welcome to join if you like :)

Reply:
Oh, I got it:)

Thank you for the invitation to the iceoryx developer meetup! I'm very interested in joining, but I'm not sure if I'll be near a computer at that time. Is this a regularly scheduled event?

Reply:
Yes, it is once a month. Currently the first Thursday at 5pm CET. Theoretically it should show up in the [ROS calendar](https://calendar.google.com/calendar/u/0/embed?src=agf3kajirket8khktupm9go748@group.calendar.google.com) but it seems it is currently removed.

Reply:
See #[af61c72](https://github.com/eclipse-iceoryx/iceoryx/pull/2069/commits/af61c72e066778b0ab153971299baa25778b6e29).